snapshot/snapshotter.go

```diff
@@ -21,131 +30,709 @@ type Mountable interface {
+			filteredChains = append(filteredChains, filteredChain)
+		}
+	}
+	applyChains = filteredChains
+
+	// Calculate the key we should use for the new snapshot via the hash of the keys that will be applied to
+	// create the merge. If there's already a snapshot with this key, then we can just re-use it rather than
+	// create a duplicate of it.
+	key := (&big.Int{}).SetBytes(digester.Sum(nil)).Text(36)[:36]
+	if info, err := sn.Stat(ctx, key); err == nil {
+		if info.Kind == snapshots.KindCommitted {
+			return key, nil
+		}
+		bklog.G(ctx).Debugf("snapshot %q is in invalid state %s, not reusing for merge", key, info.Kind.String())
+		key = identity.NewID()
+	} else if !errdefs.IsNotFound(err) {
+		return "", err
+	}
+
+	// Create the merge in a flightcontrol group for the key, ensuring parallel equivalent calls get de-duped
+	_, err := sn.mergeG.Do(ctx, key, func(ctx context.Context) (interface{}, error) {
+		// Make a new empty snapshot which will be merged into
+		prepareKey := identity.NewID()
+		if _, err := sn.snapshotter.Prepare(ctx, prepareKey, ""); err != nil {
+			return nil, errors.Wrapf(err, "failed to prepare %q", key)
+		}
+		applyMountable, err := sn.Mounts(ctx, prepareKey)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get mounts of %q", key)
+		}
+		applyMounts, unmount, err := applyMountable.Mount()
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to mount %q", key)
+		}
+
+		tempLeaseCtx, done, err := leaseutil.WithLease(ctx, sn.lm, leaseutil.MakeTemporary)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create temporary lease for view mounts during merge")
+		}
+		defer done(context.TODO())
+
+		// externalHardlinks keeps track of which inodes have been hard-linked between snapshots (which is
+		//	enabled when sn.useHardlinks is set to true)
+		externalHardlinks := make(map[uint64]struct{})
+
+		// Iterate over (lower, upper) pairs in each applyChain, calculating their diffs and applying each
+		// one to the mount the chains had to be constructed in reverse order (child->parent), so iterate
+		// in reverse.
+		for i := range applyChains {
+			chain := applyChains[len(applyChains)-1-i]
+			for j := range chain[:len(chain)-1] {
+				lower := chain[len(chain)-1-j]
+				upper := chain[len(chain)-2-j]
+
+				var lowerMounts []mount.Mount
+				if lower != "" {
+					viewID := identity.NewID()
+					var err error
+					lowerMounts, err = sn.snapshotter.View(tempLeaseCtx, viewID, lower)
+					if err != nil {
+						return nil, errors.Wrapf(err, "failed to get mounts of lower %q", lower)
+					}
+				}
+
+				viewID := identity.NewID()
+				upperMounts, err := sn.snapshotter.View(tempLeaseCtx, viewID, upper)
+				if err != nil {
+					return nil, errors.Wrapf(err, "failed to get mounts of upper %q", upper)
+				}
+
+				err = diffApply(ctx, lowerMounts, upperMounts, applyMounts, sn.useHardlinks, externalHardlinks)
```

Comment:
> Is this diffing every layer so that the Remote descriptor length are equal, no matter what internal implementation?

It's diffing so that we ensure the deletions of each layer are included when applying the layer to the merge. The current implementation always includes deletions with merges. The next PR after this will be support for conflict detection which is what will enable us to optimize merges that don't have deletion conflicts and thus optimize this step further by not needing to run the differ necessarily.

This doesn't relate to the Remote. Actually, as of right now when refs are turned into a Remote, the snapshots for their merges aren't included at all; instead the remote just consists of snapshot layers that get merged, not the merges themselves. This is probably a good example of something that would benefit from a diagram in docs; I'll be sure to include that when updating with those types of docs.

> Shouldn't it then use the actual differ, save blob and extract it if it exists?

We could do that, but it's massively inefficient relative to this approach where we just directly apply diffs and use hardlinks when possible. If we used the content store as a middle man we'd no longer be able to use hardlinks to save disk space + time copying, we'd be creating tons of duplicated files in the content store that may never be needed other than here, we'd be spending a lot of extra time exporting to the content store just to extract back into the snapshotter, etc. So while we could technically view this as an optimization, IMO it's significant enough that it's worth doing right away, especially if we end up using this approach to implement overlay merges too as I suggested in my other recent comment.